nix/ocaml.nix

```diff
@@ -85,16 +91,150 @@ let
+    commitShort = builtins.substring 0 8 commit;
+    cmdLineTest = ''
+      mina --version
+      mv _build/default/src/test/command_line_tests/command_line_tests.exe tests.exe
+      chmod +x tests.exe
+      export TMPDIR=tmp # to align with janestreet core library
+      mkdir -p $TMPDIR
+      export MINA_LIBP2P_PASS="naughty blue worm"
+      export MINA_PRIVKEY_PASS="naughty blue worm"
+      export MINA_KEYS_PATH=genesis_ledgers
+      mkdir -p $MINA_KEYS_PATH
+      echo '{"ledger":{"accounts":[]}}' > $MINA_KEYS_PATH/config_${commitShort}.json
+      ./tests.exe --mina-path mina
+    '';
+  in [
+    (dune-nix.testWithVm { } "mina_net2" [ pkgs.libp2p_helper ])
+    (dune-nix.testWithVm { } "__src-lib-mina_net2-tests__"
+      [ pkgs.libp2p_helper ])
+    (self:
+      dune-nix.testWithVm' cmdLineTest { } "__src-test-command_line_tests__" [
+        self.pkgs.cli
+        pkgs.libp2p_helper
+      ] self)
+    (dune-nix.testWithVm { } "__src-lib-staged_ledger-test__" [ ])
+  ];
+  granularCustom = _: super:
+    let
+      makefileTest = pkg:
+        let src = info.pseudoPackages."${pkg}";
+        in dune-nix.makefileTest ./.. pkg src;
+      marlinPlonkStubs = {
+        MARLIN_PLONK_STUBS = "${pkgs.kimchi_bindings_stubs}";
+      };
+      childProcessesTester = pkgs.writeShellScriptBin "mina-tester.sh"
+        (builtins.readFile ../src/lib/child_processes/tester.sh);
+      withLibp2pHelper = (s: {
+        nativeBuildInputs = s.nativeBuildInputs ++ [ pkgs.libp2p_helper ];
+      });
+    in {
+      pkgs.mina_version = super.pkgs.mina_version.overrideAttrs {
+        MINA_COMMIT_SHA1 = inputs.self.sourceInfo.rev or "<dirty>";
+      };
+      pkgs.kimchi_bindings =
+        super.pkgs.kimchi_bindings.overrideAttrs marlinPlonkStubs;
+      pkgs.kimchi_types =
+        super.pkgs.kimchi_types.overrideAttrs marlinPlonkStubs;
+      pkgs.pasta_bindings =
+        super.pkgs.pasta_bindings.overrideAttrs marlinPlonkStubs;
+      pkgs.libp2p_ipc = super.pkgs.libp2p_ipc.overrideAttrs (s: {
+        GO_CAPNP_STD = "${pkgs.go-capnproto2.src}/std";
+        nativeBuildInputs = s.nativeBuildInputs ++ [ pkgs.capnproto ];
+      });
+      pkgs.bindings_js = super.pkgs.bindings_js.overrideAttrs {
+        PLONK_WASM_NODEJS = "${pkgs.plonk_wasm}/nodejs";
+        PLONK_WASM_WEB = "${pkgs.plonk_wasm}/web";
+      };
+      files.src-lib-crypto-kimchi_bindings-js-node_js =
+        super.files.src-lib-crypto-kimchi_bindings-js-node_js.overrideAttrs {
+          PLONK_WASM_NODEJS = "${pkgs.plonk_wasm}/nodejs";
+        };
+      files.src-lib-crypto-kimchi_bindings-js-web =
+        super.files.src-lib-crypto-kimchi_bindings-js-web.overrideAttrs {
+          PLONK_WASM_WEB = "${pkgs.plonk_wasm}/web";
+        };
+      pkgs.__src-lib-ppx_mina-tests__ =
+        makefileTest "__src-lib-ppx_mina-tests__" super;
+      pkgs.__src-lib-ppx_version-test__ =
+        makefileTest "__src-lib-ppx_version-test__" super;
```

Comment:
I'm a bit worried that these declarations are so explicit. I'd assume there would be a script that auto-generates them, or at least tells if you if a new package you created is not explicitly declared in this file.

Reply:
Not exactly this way. Absolute majority of tests are autodetected. Anything that `dune test` runs will be discovered automatically.

For some of these tests we need to specify external dependencies. E.g. for `mina_lib` tests we need `libp2p_helper` compiled from Go code. With `make build` we'd inject it into environment for all the packages, in nix wrapping I decided to make it more delicate: make only relevant tests/modules dependent on an external dependency, not every single package. My approach has a benefit of say `mina_wire_types` tests not requiring `libp2p_helper` to be compiled/executed.

In this case though it's different. Both `ppx_mina` and `ppx_version` have their tests turned off the execution via `dune test`, because `dune test` interface wasn't flexible enough (sometimes we want to test that some test code **doesn't** compile with a ppx, this is not easy to do from within `dune test`). For this reason tests for these packages are overridden with explicit `makefileTest` scenario.